pitop/robotics/tilt_roll_head_controller.py

```diff
@@ -25,32 +27,89 @@ def __init__(self, servo_roll_port="S0", servo_tilt_port="S3", name="head"):
                                   )
 
         Stateful.__init__(self, children=['_roll_servo', '_tilt_servo'])
-        Recreatable.__init__(self, config_dict={'servo_roll_port': servo_roll_port, 'servo_tilt_port': servo_tilt_port, 'name': name})
+        Recreatable.__init__(self, config_dict={'servo_roll_port': servo_roll_port,
+                                                'servo_tilt_port': servo_tilt_port,
+                                                'name': name}
+                             )
 
     @property
-    def roll(self):
+    def roll(self) -> ServoMotor:
         return self._roll_servo
 
     @property
-    def tilt(self):
+    def tilt(self) -> ServoMotor:
         return self._tilt_servo
 
-    def stop(self):
+    def stop(self) -> None:
         self.tilt.sweep(speed=0)
         self.roll.sweep(speed=0)
 
-    def track_head_angle(self, angle, flipped=True):
+    def shake(self, times: int = 4, angle: int = 5, speed: int = 100, block: bool = True) -> None:
+        """Shakes the head by rotating the roll servo back and forth between
+        :data:`-angle` and :data:`+angle` from current angle position.
+
+        :param int times: Number of times to shake head
+        :param int angle: angle in degrees either side of current angle to rotate servo by
+        :param int speed: servo speed in deg/s
+        :param bool block: if `True`, function call will block program execution until finished
+        :return: None
+        """
+        self.__start_servo_oscillation(servo=self._roll_servo, times=times, angle=angle, speed=speed, block=block)
+
+    def nod(self, times: int = 4, angle: int = 5, speed: int = 100, block: bool = True) -> None:
+        """Nod the head by rotating the tilt servo back and forth between
+        :data:`-angle` and :data:`+angle` from current angle position.
+
+        :param int times: Number of times to nod head
+        :param int angle: angle in degrees either side of current angle to rotate servo by
+        :param int speed: servo speed in deg/s
+        :param bool block: if `True`, function call will block program execution until finished
+        :return: None
+        """
+        self.__start_servo_oscillation(servo=self._tilt_servo, times=times, angle=angle, speed=speed, block=block)
+
+    def __start_servo_oscillation(self, servo: ServoMotor, times: int, angle: int, speed: int, block: bool) -> None:
+        if block:
+            self.__oscillate_servo(servo=servo, times=times, angle=angle, speed=speed)
+        else:
+            Thread(target=self.__oscillate_servo,
+                   kwargs={"servo": servo, "times": times, "angle": angle, "speed": speed},
+                   daemon=True).start()
```

Comment:
what should be the behavior when 2 non-blocking movements are requested sequentially? my guess would be that the last movement is the one that gets executed...
anyway, I think we should limit the amount of threads the class can create and keep track of it, so you can "kill it" if this scenario comes up!

Reply:
@jcapona good catch - I've implemented this now let me know what you think